### PR TITLE
[FLINK-24932] Support build FRocksDB Java on Apple silicon

### PR DESCRIPTION
**Pull request by Myasuka**
This PR selected 4 necessary commits to enable building FRocksDB Java on Apple silicon.

### CODE REVIEW

**Comment by Myasuka** [comment]
@carp84 would you please take a look at this PR?
Since all necessary work to support building on apple silicon has been supported on c++ side in RocksDB-6.20.3, we only need to cherry-pick java side work to ensure it could build binaries on macos. And since we only need changes on java side, we do not need to update binaries on other platforms.

**Comment by Myasuka** [comment]
Apart from asking resources within Flink community, @adamretter , who is responsible for RocksJava in RocksDB community, would you please help to review this PR? 
This PR only includes picked changes from the RocksDB community to support running Flink forked RocksDB-6.20.3 on apple silicon chips.

**Comment by adamretter** [comment]
@Myasuka I'm not really sure what you would like me to contribute? I'm mostly just a little responsible for Java and multi-arch in RocksDB. Can you tell me how this is related?

**Comment by Myasuka** [comment]
@adamretter Thanks for your response. 
This repo is used as the Flink community forked RocksDB. We want to support RocksDB on apple silicon chips in the coming flink-1.16 release. However, Flink community lacks resources to help review RocksDB-related code currently, and since commits in this PR mainly reviewed or created by you, I think you're certainly cable of reviewing this.

The reasons why we did not bump to RocksDB-6.29.5:
1. We found a bit of performance regression in our Flink benchmark compared with RocksDB-6.20.3 with RocksDB-6.29.5.
2. We just want to avoid generating too many binaries on different platforms (we always suffer from this, I'm not sure whether you could share experiences here), and keep the code at RocksDB-6.20.3 and only cherry-pick changes on building and java side could help this. Thus, we only need to publish the jar package with updated java classes and binary `.so` files on mac-os platform.

Hope you could take a look at this PR, thanks!

**Comment by adamretter** [comment]
@Myasuka thank you, it's very kind of you to ask. However, I am afraid that I don't have enough time to also contribute to Flink.

**Comment by Myasuka** [comment]
@adamretter Thanks for your response. The feature freeze data of flink-1.16 is 2022-08-09, and you can still feel free to review this PR before that time.

This PR consists of three commits from the RocksDB community:
1. https://github.com/facebook/rocksdb/pull/9254 
2. https://github.com/facebook/rocksdb/pull/9662
3. https://github.com/facebook/rocksdb/pull/9374
The other two minor commits are introduced to keep zlib version as 1.2.11 and resolve the broken code format.

**Comment by fredia** [comment]
Hi @Myasuka, I opened circleci, can you submit this PR again to trigger tests?

File: .circleci/config.yml
Comment (rkhachatryan) [inline .circleci/config.yml:565]: NIT: double-quote `$BASH_ENV`?

File: .circleci/config.yml
Comment (rkhachatryan) [inline .circleci/config.yml:661]: Do you think it makes sense to create a ticket to re-enable it later (and reference the ticket here)?
ditto: `build-linux-gcc-4_8-no_test_run`, `build-cmake-mingw`

File: Makefile
Comment (rkhachatryan) [inline Makefile:316]: I'm wondering whether `filter-out` can remove too much here. Shouldn't it be replaced with `filter`?

File: Makefile
Comment (rkhachatryan) [inline Makefile]: Shouldn't it be `https://zlib.net/fossils/`?

**Review by rkhachatryan — APPROVED** [review]
Thanks a lot for the PR @Myasuka ,
I have only minor comments, otherwise LGTM given that the build succeeds.

**Comment by fredia** [comment]
@Myasuka , because we use the [free plan of circleCI](https://circleci.com/pricing/), we can only use S⁠–⁠L size resource size, could you please update the `.circleci/config.yaml`, and due to the limit of [memory](https://app.circleci.com/pipelines/github/ververica/frocksdb/57/workflows/2f1ab659-ff66-4d02-9ee4-f610fdaa54e5/jobs/1873) and [space](https://app.circleci.com/pipelines/github/ververica/frocksdb/57/workflows/3edf0acc-3364-4201-b7ee-fdaa742c1aae/jobs/1865), some of tests can't be finished, could we ignore these tests?

File: Makefile
Comment (Myasuka) [reply Makefile]: Actually, using `web.archive.org` to download the legacy version is referred from RocksDB community. Thanks for the information on this link!

File: .circleci/config.yml
Comment (Myasuka) [reply .circleci/config.yml:661]: I think this could be done once we make FLINK-24932 merged with the current solution.

File: .circleci/config.yml
Comment (fredia) [inline .circleci/config.yml:140]: `build-macos` failed because of `no output time`, could you please [increase timeout](https://support.circleci.com/hc/en-us/articles/360007188574-Build-has-Hit-Timeout-Limit#:~:text=CircleCI%20has%20a%20built-in,will%20be%20canceled%20and%20stopped.) here:
```
- run: 
    no_output_timeout: 30m
```

**Comment by fredia** [comment]
@Myasuka Thanks for the update, the `build-macos` failed because of `no output time`, could you please [increase timeout](https://support.circleci.com/hc/en-us/articles/360007188574-Build-has-Hit-Timeout-Limit#:~:text=CircleCI%20has%20a%20built-in,will%20be%20canceled%20and%20stopped.) at `build-macos` job:
```
- run: 
    no_output_timeout: 30m
```

**Pull request merged**
